r2i/onnxrt/engine.cc

```diff
@@ -0,0 +1,281 @@
+               "The provided frame is not an onnxrt frame");
+    return nullptr;
+  }
+
+  std::shared_ptr<Ort::Session> session = this->model->GetOnnxrtSession();
+  if (!session) {
+    error.Set(RuntimeError::Code::NULL_PARAMETER,
+              "Received null onnxrt session pointer");
+    return nullptr;
+  }
+
+  if (session->GetInputCount() > 1) {
+    error.Set(RuntimeError::Code::INCOMPATIBLE_MODEL,
+              "Number of inputs in the model is greater than 1, this is not supported");
+    return nullptr;
+  }
+
+  // Warning: We support batches of size 1 and models with
+  // 1 input only.
+  try {
+    input_node_dims = session->GetInputTypeInfo(
+                        0).GetTensorTypeAndShapeInfo().GetShape();
+  }
+
+  catch (std::exception &excep) {
+    error.Set(RuntimeError::Code::FRAMEWORK_ERROR, excep.what());
+    return nullptr;
+  }
+
+  frame_format = frame->GetFormat();
+  frame_channels = frame_format.GetNumPlanes();
+  frame_height = frame->GetHeight();
+  frame_width = frame->GetWidth();
+  input_image_size = frame_height * frame_width * frame_channels;
+
+  error = this->ValidateInputTensorShape(frame_channels, frame_height,
+                                         frame_width, input_node_dims);
+  if (error.IsError ()) {
+    return nullptr;
+  }
+
+  // Score model with input tensor, get back Prediction set with pointer
+  // of the output tensor result.
+  // Note that this implementation only supports 1 input and 1 output models.
+  error = this->ScoreModel(session, frame, input_image_size, input_node_dims,
+                           prediction);
+  if (error.IsError ()) {
+    return nullptr;
+  }
+
+  return prediction;
+}
+
+RuntimeError Engine::ValidateInputTensorShape (int channels, int height,
+    int width, std::vector<int64_t> input_dims) {
+  RuntimeError error;
+
+  // We only support 1 batch
+  if (1 != input_dims.at(0)) {
+    error.Set (RuntimeError::Code::INVALID_FRAMEWORK_PARAMETER,
+               "We only support a batch of 1 image(s) in our frames");
+    return error;
+  }
+
+  // Check that channels match
+  if (channels != input_dims.at(1)) {
+    std::string error_msg;
+    error_msg = "Channels per image:" + std::to_string(channels) +
+                ", needs to be equal to model input channels:" +
+                std::to_string(input_dims.at(1));
+    error.Set (RuntimeError::Code::INVALID_FRAMEWORK_PARAMETER, error_msg);
+    return error;
+  }
+
+  // Check that heights match
+  if (height != input_dims.at(2)) {
+    error.Set (RuntimeError::Code::INVALID_FRAMEWORK_PARAMETER,
+               "Unsupported image height");
+    return error;
+  }
+
+  // Check that widths match
+  if (width != input_dims.at(3)) {
+    error.Set (RuntimeError::Code::INVALID_FRAMEWORK_PARAMETER,
+               "Unsupported image width");
+    return error;
+  }
+
+  return error;
+}
+
+RuntimeError Engine::ScoreModel (std::shared_ptr<Ort::Session> session,
```

Comment:
This session can be a private member of this engine class

Reply:
Agree!